### ISSUE TITLE
Very high CPU load and runtime when http code 409 occurs during gravity update

### ISSUE THREAD

**Issue by thomasmerz** [pi-hole/docker-pi-hole#1334]
This is a: **Bug**  

## Details
When updating gravity I observed that for some reasons a http code 409 for one/some blocklists may occur. When this occurs gravity update seems to be stuck for a long time and CPU is running to the limits.

Please search for "(409)" in details below:
<details><pre>
09:26 # docker exec -it pihole pihole -g | /usr/bin/ts
Mar 21 09:27:10   [i] Neutrino emissions detected...
  [✓] Pulling blocklist source list into range
Mar 21 09:27:10
  [✓] Preparing new gravity database
Mar 21 09:27:10   [i] Using libz compression
Mar 21 09:27:10
Mar 21 09:27:10   [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
Mar 21 09:27:12   [i] Imported 189926 domains, ignoring 1 non-domain entries
Mar 21 09:27:12       Sample of non-domain entries:
Mar 21 09:27:12         - 0.0.0.0
Mar 21 09:27:12   [i] List stayed unchanged
Mar 21 09:27:12
Mar 21 09:27:12   [i] Target: https://someonewhocares.org/hosts/zero/hosts
  [✓] Status: No changes detected
Mar 21 09:27:13   [i] Imported 11395 domains
Mar 21 09:27:13
Mar 21 09:27:13   [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/extensions/fakenews/hosts
  [✓] Status: Retrieval successful
Mar 21 09:27:14   [i] Imported 2196 domains
Mar 21 09:27:14   [i] List stayed unchanged
Mar 21 09:27:14
Mar 21 09:27:14   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/crypto.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:14   [i] Imported 23786 domains
Mar 21 09:27:14   [i] List stayed unchanged
Mar 21 09:27:14
Mar 21 09:27:14   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/drugs.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:15   [i] Imported 26588 domains
Mar 21 09:27:15   [i] List stayed unchanged
Mar 21 09:27:15
Mar 21 09:27:15   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/fraud.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:18   [i] Imported 196085 domains
Mar 21 09:27:18   [i] List stayed unchanged
Mar 21 09:27:18
Mar 21 09:27:18   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/phishing.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:21   [i] Imported 190235 domains
Mar 21 09:27:21   [i] List stayed unchanged
Mar 21 09:27:21
Mar 21 09:27:21   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/gambling.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:22   [i] Imported 2499 domains
Mar 21 09:27:22   [i] List stayed unchanged
Mar 21 09:27:22
Mar 21 09:27:22   [i] Target: https://raw.githubusercontent.com/blocklistproject/Lists/master/piracy.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:22   [i] Imported 2134 domains
Mar 21 09:27:22   [i] List stayed unchanged
Mar 21 09:27:22
Mar 21 09:27:22   [i] Target: https://raw.githubusercontent.com/d43m0nhLInt3r/socialblocklists/master/MobileAppAds/appadsblocklist.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:22   [i] Imported 767 domains
Mar 21 09:27:22   [i] List stayed unchanged
Mar 21 09:27:22
Mar 21 09:27:22   [i] Target: https://dbl.oisd.nl/
  [✓] Status: No changes detected
Mar 21 09:27:27   [i] Imported 851695 domains
Mar 21 09:27:27
Mar 21 09:27:27   [i] Target: https://zerodot1.gitlab.io/CoinBlockerLists/hosts
  [✓] Status: No changes detected
Mar 21 09:27:29   [i] Imported 283969 domains
Mar 21 09:27:29
Mar 21 09:27:29   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/Corona-Blocklist
  [✓] Status: Retrieval successful
Mar 21 09:27:34   [i] Imported 441936 domains
Mar 21 09:27:34   [i] List stayed unchanged
Mar 21 09:27:34
Mar 21 09:27:34   [i] Target: https://www.technoy.de/lists/FireTVAds.txt
  [✓] Status: No changes detected
Mar 21 09:27:35   [i] Imported 20 domains
Mar 21 09:27:35
Mar 21 09:27:35   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/Phishing-Angriffe
  [✓] Status: Retrieval successful
Mar 21 09:27:39   [i] Imported 476668 domains
Mar 21 09:27:39   [i] List stayed unchanged
Mar 21 09:27:39
Mar 21 09:27:39   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/Streaming
  [✓] Status: Retrieval successful
Mar 21 09:27:40   [i] Imported 4186 domains
Mar 21 09:27:40   [i] List stayed unchanged
Mar 21 09:27:40
Mar 21 09:27:40   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/malware
  [✓] Status: Retrieval successful
Mar 21 09:27:41   [i] Imported 91478 domains
Mar 21 09:27:41   [i] List stayed unchanged
Mar 21 09:27:41
Mar 21 09:27:41   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/child-protection
  [✓] Status: Retrieval successful
Mar 21 09:27:46   [i] Imported 414050 domains
Mar 21 09:27:46   [i] List stayed unchanged
Mar 21 09:27:46
Mar 21 09:27:46   [i] Target: https://raw.githubusercontent.com/crazy-max/WindowsSpyBlocker/master/data/hosts/spy.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:47   [i] Imported 347 domains
Mar 21 09:27:47   [i] List stayed unchanged
Mar 21 09:27:47
Mar 21 09:27:47   [i] Target: https://raw.githubusercontent.com/crazy-max/WindowsSpyBlocker/master/data/hosts/spy_v6.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:47   [i] Imported 347 domains
Mar 21 09:27:47   [i] List stayed unchanged
Mar 21 09:27:47
Mar 21 09:27:47   [i] Target: https://raw.githubusercontent.com/lightswitch05/hosts/master/docs/lists/ads-and-tracking-extended.txt
  [✓] Status: Retrieval successful
Mar 21 09:27:53   [i] Imported 431088 domains
Mar 21 09:27:53   [i] List has been updated
Mar 21 09:27:53
Mar 21 09:27:53   [i] Target: https://dbl.oisd.nl/nsfw/
  [✗] Status: https://dbl.oisd.nl/nsfw/ (409)
Mar 21 09:27:53   [✗] List download failed: using previously cached list
Mar 21 09:36:52
Mar 21 09:36:54
Mar 21 09:37:14   [i] Imported 0 domains, ignoring 328601 non-domain entries
Mar 21 09:37:14       Sample of non-domain entries:
Mar 21 09:37:14         - 2023-03-16t00:43:32+0000
Mar 21 09:37:14         - 202303160043
Mar 21 09:37:14         - frequency)
Mar 21 09:37:14         - nsfw
Mar 21 09:37:14         - plus]
Mar 21 09:37:15
Mar 21 09:37:15   [i] Target: https://raw.githubusercontent.com/lightswitch05/hosts/master/docs/lists/hate-and-junk-extended.txt
  [✓] Status: Retrieval successful
Mar 21 09:37:15   [i] Imported 1679 domains
Mar 21 09:37:15   [i] List has been updated
Mar 21 09:37:16
Mar 21 09:37:16   [i] Target: https://raw.githubusercontent.com/lightswitch05/hosts/master/docs/lists/tracking-aggressive-extended.txt
  [✓] Status: Retrieval successful
Mar 21 09:37:19   [i] Imported 171732 domains
Mar 21 09:37:19   [i] List has been updated
Mar 21 09:37:19
Mar 21 09:37:19   [i] Target: https://raw.githubusercontent.com/RPiList/specials/master/Blocklisten/spam.mails
  [✓] Status: Retrieval successful
Mar 21 09:37:26   [i] Imported 539404 domains
Mar 21 09:37:26   [i] List stayed unchanged
Mar 21 09:37:26
Mar 21 09:37:26   [i] Target: https://raw.githubusercontent.com/Perflyst/PiHoleBlocklist/master/android-tracking.txt
  [✓] Status: Retrieval successful
Mar 21 09:37:27   [i] Imported 80 domains
Mar 21 09:37:27   [i] List stayed unchanged
Mar 21 09:37:27
Mar 21 09:37:27   [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/data/minecraft-hosts/hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:27   [i] Imported 6 domains
Mar 21 09:37:27   [i] List stayed unchanged
Mar 21 09:37:27
Mar 21 09:37:27   [i] Target: https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/snuff-hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:28   [i] Imported 23 domains
Mar 21 09:37:29   [i] List stayed unchanged
Mar 21 09:37:29
Mar 21 09:37:29   [i] Target: https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/pornography-hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:30   [i] Imported 19670 domains
Mar 21 09:37:30   [i] List stayed unchanged
Mar 21 09:37:30
Mar 21 09:37:30   [i] Target: https://raw.githubusercontent.com/tiuxo/hosts/master/porn
  [✓] Status: Retrieval successful
Mar 21 09:37:30   [i] Imported 369 domains
Mar 21 09:37:30   [i] List stayed unchanged
Mar 21 09:37:31
Mar 21 09:37:31   [i] Target: https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/gambling-hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:31   [i] Imported 2571 domains
Mar 21 09:37:31   [i] List stayed unchanged
Mar 21 09:37:32
Mar 21 09:37:32   [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/extensions/porn/clefspeare13/hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:33   [i] Imported 30787 domains
Mar 21 09:37:33   [i] List stayed unchanged
Mar 21 09:37:33
Mar 21 09:37:33   [i] Target: https://mypdns.org/my-privacy-dns/matrix/-/raw/master/source/porn_filters/explicit_content/domains.list
  [✗] Status: Connection Refused
Mar 21 09:37:33   [✗] List download failed: using previously cached list
Mar 21 09:37:34   [i] Imported 5630 domains
Mar 21 09:37:34
Mar 21 09:37:34   [i] Target: https://mypdns.org/my-privacy-dns/matrix/-/raw/master/source/porn_filters/explicit_content/hosts.list
  [✗] Status: Connection Refused
Mar 21 09:37:34   [✗] List download failed: using previously cached list
Mar 21 09:37:35   [i] Imported 56712 domains
Mar 21 09:37:35
Mar 21 09:37:35   [i] Target: https://raw.githubusercontent.com/liamengland1/mischosts/master/apple-telemetry
  [✓] Status: Retrieval successful
Mar 21 09:37:36   [i] Imported 234 domains
Mar 21 09:37:36   [i] List stayed unchanged
Mar 21 09:37:36
Mar 21 09:37:36   [i] Target: https://raw.githubusercontent.com/liamengland1/mischosts/master/microsoft-telemetry
  [✓] Status: Retrieval successful
Mar 21 09:37:37   [i] Imported 916 domains
Mar 21 09:37:37   [i] List stayed unchanged
Mar 21 09:37:37
Mar 21 09:37:37   [i] Target: https://raw.githubusercontent.com/liamengland1/mischosts/master/tiktok-hosts
  [✓] Status: Retrieval successful
Mar 21 09:37:37   [i] Imported 184 domains
Mar 21 09:37:37   [i] List stayed unchanged
Mar 21 09:37:38
Mar 21 09:37:38   [i] Target: https://hosts.tweedge.net/malicious.txt
  [✓] Status: Retrieval successful
Mar 21 09:37:39   [i] Imported 2325 domains
Mar 21 09:37:39   [i] List has been updated
Mar 21 09:37:39
Mar 21 09:37:39   [i] Target: https://pgl.yoyo.org/as/serverlist.php?showintro=0;hostformat=hosts
  [✓] Status: No changes detected
Mar 21 09:37:39   [i] Imported 3757 domains
Mar 21 09:37:40
  [✓] Creating new gravity databases
  [✓] Storing downloaded domains in new gravity database
  [✓] Building tree
  [✓] Swapping databases
Mar 21 09:38:17   [i] Number of gravity domains: 4477474 (3900774 unique domains)
Mar 21 09:38:17   [i] Number of exact blacklisted domains: 19
Mar 21 09:38:17   [i] Number of regex blacklist filters: 23
Mar 21 09:38:17   [i] Number of exact whitelisted domains: 12
Mar 21 09:38:17   [i] Number of regex whitelist filters: 3
  [✓] Flushing DNS cache
  [✓] Cleaning up stray matter
Mar 21 09:38:18
Mar 21 09:38:18   [✓] FTL is listening on port 53
Mar 21 09:38:18      [✓] UDP (IPv4)
Mar 21 09:38:18      [✓] TCP (IPv4)
Mar 21 09:38:18      [✓] UDP (IPv6)
Mar 21 09:38:18      [✓] TCP (IPv6)
Mar 21 09:38:18
Mar 21 09:38:18   [✓] Pi-hole blocking is enabled
🇫🇮  ✔ ~ [master|✔]
09:38 #
</pre></details>

This is what `sar` says on my small single-CPU-cloudserver during gravity update:
```
🇫🇮  ✔ ~ [master|✔]
09:40 # sar -s 08:45
Linux 5.15.0-52-generic (ubuntu-cx11-03) 	03/21/2023 	_x86_64_	(1 CPU)

08:45:02 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
08:55:01 AM     all      9.70      2.77     11.08      0.05      0.00     76.39
09:05:01 AM     all     13.38      2.78     12.27      0.29      0.00     71.28
09:15:01 AM     all     10.77      2.85     11.52      0.24      0.00     74.62
09:25:01 AM     all     10.55      3.47     11.26      0.08      0.00     74.64
09:35:01 AM     all     73.75      1.82      8.42      0.20      0.00     15.81
09:45:01 AM     all     34.27      2.34     11.71      0.31      0.00     51.37
Average:        all     25.53      2.67     11.04      0.19      0.00     60.57
🇫🇮  ✔ ~ [master|✔]
09:46 #
```

But there's also an abnormal high CPU load/usage on my four-CPU linux desktop during gravity update :
```
🦎🖥  thomas@merz-nimbus:~ [1/2684]
09:15 $ sar -s 01:30:01 -e 02:25:01
Linux 5.14.21-150400.24.46-default (merz-nimbus) 	03/21/2023 	_x86_64_	(4 CPU)

01:30:01 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
01:40:01 AM     all      2.41      1.51      3.44      0.08      0.00     92.56
01:50:01 AM     all      2.51      0.90      3.39      0.07      0.00     93.13
02:00:01 AM     all     16.20      0.95      3.40      0.05      0.00     79.39
02:10:01 AM     all      8.34      1.03      3.32      0.10      0.00     87.21
02:20:01 AM     all      2.47      0.82      3.36      0.04      0.00     93.31
Average:        all      6.39      1.04      3.38      0.07      0.00     89.12
🦎🖥  thomas@merz-nimbus:~ [1/2685]
```

## Related Issues
- [x] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 
- [ ] I reported this 409 at the blocklist's owner repo: https://github.com/sjhgvr/oisd/issues/1

## How to reproduce the issue 

1. Environment data
  * Operating System: differerent Linux systems running Pi-hole in Docker Container
  * Hardware: PC
  * Kernel Architecture: x86/amd64
  * Hardware architecture: x86

2. reproducable at any time (given that a 409 will occur what is mostly the case) by executing `docker exec -it pihole pihole -g | /usr/bin/ts`

## These common fixes didn't work for my issue
- [ ] I have tried removing/destroying my container, and re-creating a new container
- [ ] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [ ] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [ ] I have tried running without my volume data mounts to eliminate volumes as the cause

**Comment by PromoFaux** [comment]
Is this specific to Docker, or can it be produced on bare metal, too? There is nothing special about how Docker handles gravity, so suspect the latter.

That said, I wonder if the high CPU usage can be attributed to:

> Mar 21 09:37:14   [i] Imported 0 domains, ignoring 328601 non-domain entries

There is nothing on that list that Pi-hole can import in any case, adding it to your set up gains you nothing ;) I'm assuming the list is in adblock plus format? (Difficult to tell when it is offline)

We have the following upcoming feature that will implement adblock plus style parsing:

https://github.com/pi-hole/FTL/pull/1532

**Comment by PromoFaux** [comment]
> There is nothing special about how Docker handles gravity, so suspect the latter.

That said, I am now vaguely recalling there being an issue with how things are parsed - though am struggling to find the specifics, @dschaper may know

**Comment by PromoFaux** [comment]
@thomasmerz in the meantime, if you're feeling brave can you try the `nightly` tag and see if it still shows the same symptoms? That tag has got the ABP  style changes in it

**Comment by thomasmerz** [comment]
Sorry, I don't have any Pi-hole running on bare metal…  
But by working through so many domains to be ignored this might lead to a very high imperfomant and long run 🤔 

Running with `Docker Tag nightly` and
>    Pi-hole vDev (development, v5.15.5-36-g7ea0bbb)
    FTL vDev (development, vDev-d35edd4)
    Web Interface vDev (devel, v5.18.4-32-g94448ed)

it's pretty much better ❤️ 

```
Mar 21 12:04:51   [i] Target: https://dbl.oisd.nl/nsfw/
  [✗] Status: https://dbl.oisd.nl/nsfw/ (409)
Mar 21 12:04:51   [✗] List download failed: using previously cached list
Mar 21 12:04:51   [i] List contained AdBlock Plus style domains
Mar 21 12:04:53   [i] Imported 328596 domains, ignoring 5 non-domain entries
Mar 21 12:04:53       Sample of non-domain entries:
Mar 21 12:04:53         - 2023-03-16t00:43:32+0000
Mar 21 12:04:53         - 202303160043
Mar 21 12:04:53         - frequency)
Mar 21 12:04:53         - nsfw
Mar 21 12:04:53         - plus]
```

So shall/can we close this as "will be fixed in upcoming version"?

**Comment by dschaper** [comment]
The current code filters out a small set of domains that are not helpful for blocklists (like localhost or localdomain or `::`) that was added because of a change in the default suggested list that we ship. That code adds a dozen or so loops though the entirety of a list and for some reason really slogs down Docker when large lists that contain no usable data are sourced. There's no problem on bare metal. 

We didn't bother investigating more because the whole filter is removed in the next release.

**Comment by PromoFaux** [comment]
Changes released with `2023.03.0`
